helix-core/src/modeline.rs

```diff
@@ -0,0 +1,292 @@
+const LENGTH_TO_CHECK: usize = 256;
+
+static VIM_MODELINE_REGEX: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"^(\S*\s+)?(vi|[vV]im[<=>]?\d*|ex):\s*(set?\s+)?").unwrap());
+static HELIX_MODELINE_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"^(\S*\s+)?helix:").unwrap());
+
+#[derive(Default, Debug, Eq, PartialEq)]
+pub struct Modeline {
+    language: Option<String>,
+    indent_style: Option<IndentStyle>,
+    line_ending: Option<LineEnding>,
+}
+
+impl Modeline {
+    pub fn parse(text: RopeSlice) -> Self {
+        let mut modeline = Self::default();
+
+        for line in text.lines().take(LINES_TO_CHECK).chain(
+            text.lines_at(text.len_lines())
+                .reversed()
+                .take(LINES_TO_CHECK),
+        ) {
+            // can't guarantee no extra copies, since we need to regex and
+            // regexes can't operate over chunks yet, but we can at least
+            // limit how much we potentially need to copy because modelines
+            // are typically quite short.
+            if line.len_chars() > LENGTH_TO_CHECK {
+                continue;
+            }
+            let line = Cow::<str>::from(line);
+            modeline.parse_from_line(&line);
+        }
+
+        modeline
+    }
+
+    pub fn language(&self) -> Option<&str> {
+        self.language.as_deref()
+    }
+
+    pub fn indent_style(&self) -> Option<IndentStyle> {
+        self.indent_style
+    }
+
+    pub fn line_ending(&self) -> Option<LineEnding> {
+        self.line_ending
+    }
+
+    fn parse_from_line(&mut self, line: &str) {
+        let mut saw_backslash = false;
+        let split_modeline = move |c| {
+            saw_backslash = match c {
+                ':' if !saw_backslash => return true,
+                '\\' => true,
+                _ => false,
+            };
+            c == ' ' || c == '\t'
+        };
+
+        if let Some(pos) = VIM_MODELINE_REGEX.find(line) {
+            for option in line[pos.end()..].split(split_modeline) {
+                let parts: Vec<_> = option.split('=').collect();
+                match parts[0] {
+                    "ft" | "filetype" => {
+                        if let Some(val) = parts.get(1) {
+                            self.language = Some(val.to_string());
+                        }
+                    }
+                    "sw" | "shiftwidth" => {
+                        if let Some(val) = parts.get(1).and_then(|val| val.parse().ok()) {
+                            if self.indent_style != Some(IndentStyle::Tabs) {
+                                self.indent_style = Some(IndentStyle::Spaces(val));
+                            }
+                        }
+                    }
+                    "ff" | "fileformat" => {
+                        if let Some(val) = parts.get(1) {
+                            self.line_ending = vim_ff_to_helix_line_ending(val);
+                        }
+                    }
+                    "noet" | "noexpandtab" => {
```

Comment:
I wonder if we should habdle `expandtab` without an explicit shiftwidth. I think that isn't too rare. We could simply set the indentation to the tabwidth currently set for that document.  This would match `sw=0` (which we may also want to handle) and is generally what a user would expect.

One minor csbeat in both cases: If the document ks already configured to use spaces then expandtab should do nothing. That will need special representation in the ModeLine struct but IMO it's worth it since that is pretty common.